apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -77,6 +64,231 @@ mod prepare_block {
+        ) -> Vec<TxRecord> {
+            let protocol_key = self
+                .mode
+                .get_protocol_key()
+                .expect("Validators should always have a protocol key");
+
+            let vote_extension_digest =
+                local_last_commit.and_then(|local_last_commit| {
+                    let votes = local_last_commit.votes;
+                    self.compress_vote_extensions(votes)
+                });
+            let vote_extension_digest = match vote_extension_digest {
+                Some(_) if self.storage.last_height.0 == 0 => {
+                    tracing::error!(
+                        "The genesis block should not contain vote extensions"
+                    );
+                    return vec![];
+                }
+                Some(d) => d,
+                // if no vote extensions were found, we return an empty
+                // `Vec` of protocol
+                // transactions
+                _ => return vec![],
+            };
+
+            let tx = ProtocolTxType::EthereumEvents(vote_extension_digest)
+                .sign(&protocol_key)
+                .to_bytes();
+            let tx_record = record::add(tx);
+
+            vec![tx_record]
+        }
+
+        /// Builds a batch of mempool transactions
+        fn build_mempool_txs(&mut self, txs: Vec<Vec<u8>>) -> Vec<TxRecord> {
+            // filter in half of the new txs from Tendermint, only keeping
+            // wrappers
+            let number_of_new_txs = 1 + txs.len() / 2;
+            txs.into_iter()
+                .take(number_of_new_txs)
+                .map(|tx_bytes| {
+                    if let Ok(Ok(TxType::Wrapper(_))) =
+                        Tx::try_from(tx_bytes.as_slice()).map(process_tx)
+                    {
+                        record::keep(tx_bytes)
+                    } else {
+                        record::remove(tx_bytes)
+                    }
+                })
+                .collect()
+        }
+
+        /// Builds a batch of DKG decrypted transactions
+        fn build_decrypted_txs(&mut self) -> Vec<TxRecord> {
+            // TODO: This should not be hardcoded
+            let privkey = <EllipticCurve as PairingEngine>::G2Affine::prime_subgroup_generator();
+
+            self.storage
+                .tx_queue
+                .iter()
+                .map(|tx| {
+                    Tx::from(match tx.decrypt(privkey) {
+                        Ok(tx) => DecryptedTx::Decrypted(tx),
+                        _ => DecryptedTx::Undecryptable(tx.clone()),
+                    })
+                    .to_bytes()
+                })
+                .map(record::add)
+                .collect()
+        }
+
+        /// Compresses a set of vote extensions into a single
+        /// [`VoteExtensionDigest`], whilst filtering invalid
+        /// `Signed<VoteExtension>` instances in the process
+        fn compress_vote_extensions(
```

Comment:
I would like to have a few unit tests for this function. 
 - Happy flow
 - Deduplicate multiple events from same validator
 - Filter out non-validator vote extensions
 - Filter out bad signatures from validators